File: test_apis.js
#!/usr/bin/env node
/**
 * 🚀 Testador de APIs REST - Quiz Quest
 * Execute: node test_apis.js
 */

const https = require('https');
const http = require('http');

class QuizAPITester {
    constructor(baseUrl = 'http://localhost:3000/api') {
        this.baseUrl = baseUrl;
    }

    // Função para fazer requisições HTTP
    async makeRequest(url) {
        return new Promise((resolve, reject) => {
            const module = url.startsWith('https') ? https : http;
            
            const req = module.get(url, (res) => {
                let data = '';
                
                res.on('data', (chunk) => {
                    data += chunk;
                });
                
                res.on('end', () => {
                    try {
                        const jsonData = JSON.parse(data);
                        resolve({ statusCode: res.statusCode, data: jsonData });
                    } catch (error) {
                        resolve({ statusCode: res.statusCode, data: data, error: 'JSON Parse Error' });
                    }
                });
            });
            
            req.on('error', (error) => {
                reject(error);
            });
            
            req.setTimeout(10000, () => {
                req.destroy();
                reject(new Error('Timeout'));
            });
        });
    }

    // Imprimir cabeçalho
    printHeader(title) {
        console.log('\n' + '='.repeat(60));
        console.log(`🎯 ${title}`);
        console.log('='.repeat(60));
    }

    // Testar um endpoint específico
    async testEndpoint(endpoint, description) {
        const url = `${this.baseUrl}${endpoint}`;
        console.log(`\n📡 Testando: ${description}`);
        console.log(`URL: ${url}`);
        
        try {
            const startTime = Date.now();
            const response = await this.makeRequest(url);
            const endTime = Date.now();
            
            console.log(`⏱️  Tempo de resposta: ${endTime - startTime}ms`);
            console.log(`📊 Status Code: ${response.statusCode}`);
            
            if (response.statusCode === 200 && response.data) {
                if (response.data.success) {
                    const count = response.data.data ? response.data.data.length : 0;
                    console.log(`✅ Sucesso! ${count} registros encontrados`);
                    
                    // Mostrar amostra dos dados
                    if (count > 0) {
                        console.log(`📋 Amostra do primeiro registro:`);
                        const sample = response.data.data[0];
                        const keys = Object.keys(sample).slice(0, 3);
                        keys.forEach(key => {
                            console.log(`   • ${key}: ${sample[key]}`);
                        });
                        if (Object.keys(sample).length > 3) {
                            console.log(`   ... e mais ${Object.keys(sample).length - 3} campos`);
                        }
                    }
                    
                    return { success: true, data: response.data, count };
                } else {
                    console.log(`❌ API retornou erro:`, response.data);
                    return { success: false, data: response.data, count: 0 };
                }
            } else {
                console.log(`❌ Erro HTTP ${response.statusCode}`);
                return { success: false, data: null, count: 0 };
            }
            
        } catch (error) {
            if (error.message === 'Timeout') {
                console.log('❌ Timeout - Servidor demorou para responder');
            } else if (error.code === 'ECONNREFUSED') {
                console.log('❌ Erro de conexão - Servidor não está rodando?');
            } else {
                console.log(`❌ Erro inesperado: ${error.message}`);
            }
            return { success: false, data: null, count: 0 };
        }
    }

    // Executar todos os testes
    async runAllTests() {
        this.printHeader('TESTADOR DE APIs - QUIZ QUEST');
        console.log(`🔗 Base URL: ${this.baseUrl}`);
        console.log(`⏰ Timestamp: ${new Date().toLocaleString('pt-BR')}`);
        
        const endpoints = [
            ['/quiz-results', 'Resultados dos Quizzes'],
            ['/conversion-events', 'Eventos de Conversão'],
            ['/hotmart-purchases', 'Compras Hotmart'],
            ['/utm-analytics', 'Analytics UTM'],
            ['/quiz-participants', 'Participantes do Quiz']
        ];
        
        const results = [];
        let successCount = 0;
        
        for (const [endpoint, description] of endpoints) {
            const result = await this.testEndpoint(endpoint, description);
            results.push({
                endpoint,
                description,
                success: result.success,
                count: result.count
            });
            if (result.success) successCount++;
        }
        
        // Resumo final
        this.printHeader('RESUMO DOS TESTES');
        console.log(`✅ APIs funcionando: ${successCount}/${endpoints.length}`);
        console.log(`❌ APIs com problema: ${endpoints.length - successCount}/${endpoints.length}`);
        
        console.log('\n📊 Detalhamento:');
        results.forEach(result => {
            const status = result.success ? '✅' : '❌';
            console.log(`${status} ${result.endpoint} - ${result.description} (${result.count} registros)`);
        });
        
        if (successCount === endpoints.length) {
            console.log('\n🎉 PARABÉNS! Todas as APIs estão funcionando perfeitamente!');
        } else if (successCount > 0) {
            console.log('\n⚠️  Algumas APIs estão funcionando. Verifique os erros acima.');
        } else {
            console.log('\n🚨 ATENÇÃO! Nenhuma API está funcionando. Verifique se o servidor está rodando:');
            console.log('   1. Execute: npm run dev');
            console.log('   2. Verifique se a porta 3000 está correta');
            console.log('   3. Teste manualmente: curl http://localhost:3000/api/quiz-results');
        }
        
        return results;
    }

    // Testar jornada de usuário específico
    async testUserJourney(email) {
        this.printHeader(`JORNADA DO USUÁRIO: ${email}`);
        
        const result = await this.testEndpoint(`/conversion-events/email/${email}`, `Eventos do usuário ${email}`);
        
        if (result.success && result.data && result.data.data) {
            const events = result.data.data;
            console.log('\n📈 Jornada completa encontrada:');
            events.forEach((event, index) => {
                const timestamp = event.createdAt || 'N/A';
                const eventType = event.eventType || 'N/A';
                const value = event.value || 0;
                console.log(`   ${index + 1}. ${timestamp} - ${eventType} (R$ ${value})`);
            });
        }
        
        return result;
    }

    // Exemplo de como usar as APIs em uma aplicação
    async demonstrateUsage() {
        this.printHeader('EXEMPLO DE USO DAS APIs');
        
        console.log('📝 Exemplo de como usar essas APIs em sua aplicação:');
        console.log(`
// 1. Buscar todos os resultados dos quizzes
fetch('${this.baseUrl}/quiz-results')
  .then(response => response.json())
  .then(data => {
    console.log('Resultados:', data.data);
    // Processar dados...
  });

// 2. Buscar eventos de conversão
fetch('${this.baseUrl}/conversion-events')
  .then(response => response.json())
  .then(data => {
    const events = data.data;
    console.log('Total de eventos:', events.length);
    // Criar dashboard...
  });

// 3. Buscar jornada de um usuário
fetch('${this.baseUrl}/conversion-events/email/user@example.com')
  .then(response => response.json())
  .then(data => {
    console.log('Jornada do usuário:', data.data);
    // Análise de comportamento...
  });
        `);
    }
}

// Função principal
async function main() {
    console.log('🚀 Iniciando teste das APIs REST...');
    
    // Criar instância do testador
    const tester = new QuizAPITester();
    
    // Executar todos os testes
    await tester.runAllTests();
    
    // Teste adicional: jornada de usuário
    console.log('\n' + '='.repeat(60));
    await tester.testUserJourney('user@example.com');
    
    // Demonstrar uso das APIs
    await tester.demonstrateUsage();
    
    console.log('\n🏁 Teste finalizado!');
    console.log('💡 Para visualizar os dados em um dashboard, abra: dashboard_analytics.html');
}

// Executar se for chamado diretamente
if (require.main === module) {
    main().catch(console.error);
}

module.exports = QuizAPITester;

File: test_apis.py
#!/usr/bin/env python3
"""
🚀 Script para Testar APIs REST do Quiz Quest
Execute: python test_apis.py
"""

import requests
import json
import time
from datetime import datetime

# Configuração
BASE_URL = "http://localhost:3000/api"
TIMEOUT = 10

class QuizAPITester:
    def __init__(self, base_url=BASE_URL):
        self.base_url = base_url
        self.session = requests.Session()
        self.session.timeout = TIMEOUT
        
    def print_header(self, title):
        print(f"\n{'='*60}")
        print(f"🎯 {title}")
        print(f"{'='*60}")
        
    def test_api(self, endpoint, description):
        """Testa um endpoint específico"""
        url = f"{self.base_url}{endpoint}"
        print(f"\n📡 Testando: {description}")
        print(f"URL: {url}")
        
        try:
            start_time = time.time()
            response = self.session.get(url)
            end_time = time.time()
            
            print(f"⏱️  Tempo de resposta: {(end_time - start_time)*1000:.0f}ms")
            print(f"📊 Status Code: {response.status_code}")
            
            if response.status_code == 200:
                data = response.json()
                if data.get('success'):
                    count = len(data.get('data', []))
                    print(f"✅ Sucesso! {count} registros encontrados")
                    
                    # Mostrar amostra dos dados se existirem
                    if count > 0:
                        print(f"📋 Amostra do primeiro registro:")
                        sample = data['data'][0]
                        for key, value in list(sample.items())[:3]:  # Primeiros 3 campos
                            print(f"   • {key}: {value}")
                        if len(sample.items()) > 3:
                            print(f"   ... e mais {len(sample.items()) - 3} campos")
                    
                    return True, data
                else:
                    print(f"❌ API retornou erro: {data}")
                    return False, data
            else:
                print(f"❌ Erro HTTP {response.status_code}: {response.text[:200]}")
                return False, None
                
        except requests.exceptions.ConnectionError:
            print("❌ Erro de conexão - Servidor não está rodando?")
            return False, None
        except requests.exceptions.Timeout:
            print("❌ Timeout - Servidor demorou para responder")
            return False, None
        except Exception as e:
            print(f"❌ Erro inesperado: {e}")
            return False, None
    
    def run_all_tests(self):
        """Executa todos os testes de API"""
        self.print_header("TESTADOR DE APIs - QUIZ QUEST")
        print(f"🔗 Base URL: {self.base_url}")
        print(f"⏰ Timestamp: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
        
        # Lista de endpoints para testar
        endpoints = [
            ("/quiz-results", "Resultados dos Quizzes"),
            ("/conversion-events", "Eventos de Conversão"),
            ("/hotmart-purchases", "Compras Hotmart"),
            ("/utm-analytics", "Analytics UTM"),
            ("/quiz-participants", "Participantes do Quiz"),
        ]
        
        results = []
        success_count = 0
        
        for endpoint, description in endpoints:
            success, data = self.test_api(endpoint, description)
            results.append({
                'endpoint': endpoint,
                'description': description,
                'success': success,
                'data_count': len(data.get('data', [])) if success and data else 0
            })
            if success:
                success_count += 1
        
        # Resumo final
        self.print_header("RESUMO DOS TESTES")
        print(f"✅ APIs funcionando: {success_count}/{len(endpoints)}")
        print(f"❌ APIs com problema: {len(endpoints) - success_count}/{len(endpoints)}")
        
        print(f"\n📊 Detalhamento:")
        for result in results:
            status = "✅" if result['success'] else "❌"
            print(f"{status} {result['endpoint']} - {result['description']} ({result['data_count']} registros)")
        
        if success_count == len(endpoints):
            print(f"\n🎉 PARABÉNS! Todas as APIs estão funcionando perfeitamente!")
        elif success_count > 0:
            print(f"\n⚠️  Algumas APIs estão funcionando. Verifique os erros acima.")
        else:
            print(f"\n🚨 ATENÇÃO! Nenhuma API está funcionando. Verifique se o servidor está rodando:")
            print(f"   1. Execute: npm run dev")
            print(f"   2. Verifique se a porta {self.base_url.split(':')[-1].split('/')[0]} está correta")
            print(f"   3. Teste manualmente: curl {self.base_url}/quiz-results")
        
        return results
    
    def test_specific_user(self, email):
        """Testa a jornada de um usuário específico"""
        self.print_header(f"JORNADA DO USUÁRIO: {email}")
        
        endpoint = f"/conversion-events/email/{email}"
        success, data = self.test_api(endpoint, f"Eventos do usuário {email}")
        
        if success and data.get('data'):
            events = data['data']
            print(f"\n📈 Jornada completa encontrada:")
            for i, event in enumerate(events, 1):
                timestamp = event.get('createdAt', 'N/A')
                event_type = event.get('eventType', 'N/A')
                value = event.get('value', 0)
                print(f"   {i}. {timestamp} - {event_type} (R$ {value})")
        
        return success, data

def main():
    """Função principal"""
    print("🚀 Iniciando teste das APIs REST...")
    
    # Criar instância do testador
    tester = QuizAPITester()
    
    # Executar todos os testes
    results = tester.run_all_tests()
    
    # Teste adicional: buscar jornada de um usuário exemplo
    print(f"\n" + "="*60)
    test_email = "user@example.com"
    tester.test_specific_user(test_email)
    
    print(f"\n🏁 Teste finalizado!")
    print(f"💡 Para visualizar os dados em um dashboard, abra: dashboard_analytics.html")

if __name__ == "__main__":
    main()
